src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -404,42 +404,62 @@ static bool IsRootedCallback(IntPtr pObj)
             {
                 // We are paused in the GC, so this is safe.
 #pragma warning disable CS8500 // Takes a pointer to a managed type
-                ManagedObjectWrapperHolder* holder = (ManagedObjectWrapperHolder*)&pObj;
+                WrappedObjectHolder* holder = (WrappedObjectHolder*)&pObj;
                 return holder->_wrapper->IsRooted;
 #pragma warning restore CS8500
             }
 
-            private ManagedObjectWrapper* _wrapper;
-            private object _wrappedObject;
+            public object WrappedObject { get; }
+            private readonly ManagedObjectWrapper* _wrapper;
 
-            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper, object wrappedObject)
+            public WrappedObjectHolder(object wrappedObject, ManagedObjectWrapper* wrapper)
             {
+                WrappedObject = wrappedObject;
                 _wrapper = wrapper;
-                _wrappedObject = wrappedObject;
             }
+        }
 
-            public void InitializeHandle()
+        internal unsafe class ManagedObjectWrapperHolder
+        {
+            private ManagedObjectWrapper* _wrapper;
+            private DependentHandle _wrappedObject;
+            private bool _exposed;
+
+            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper, object wrappedObject)
             {
-                IntPtr handle = RuntimeImports.RhHandleAllocRefCounted(this);
-                IntPtr prev = Interlocked.CompareExchange(ref _wrapper->HolderHandle, handle, IntPtr.Zero);
-                if (prev != IntPtr.Zero)
-                {
-                    RuntimeImports.RhHandleFree(handle);
-                }
+                _wrapper = wrapper;
+                var wrappedObjectHolder = new WrappedObjectHolder(wrappedObject, _wrapper);
+                _wrappedObject = new DependentHandle(wrappedObject, wrappedObjectHolder);
+                _wrapper->HolderHandle = RuntimeImports.RhHandleAllocRefCounted(wrappedObjectHolder);
             }
 
             public unsafe IntPtr ComIp => _wrapper->As(in ComWrappers.IID_IUnknown);
 
-            public object WrappedObject => _wrappedObject;
-
-            public uint AddRef() => _wrapper->AddRef();
+            public uint AddRef(object wrappedObject)
+            {
+                _exposed = true;
```

Comment:
I'm not sure if I need more synchronization on this variable. It is set to true on any random thread and read from the finializer thread.

Reply:
I'm just going to remove this for now. I'll revisit the issue of extra memory being used when there is a race condition inserting into the conditional weak table later.